crates/types/src/net/mod.rs

```diff
@@ -51,22 +52,61 @@ pub enum AdvertisedAddress {
     /// Unix domain socket
     #[display("unix:{}", _0.display())]
     Uds(PathBuf),
-    /// Hostname or host:port pair, or any unrecognizable string.
+    /// Hostname or host:port pair
     #[display("{}", _0)]
     Http(Uri),
 }
 
 impl FromStr for AdvertisedAddress {
-    type Err = http::uri::InvalidUri;
+    type Err = Error;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        if let Some(stripped_address) = s.strip_prefix("unix:") {
-            Ok(AdvertisedAddress::Uds(
-                stripped_address.parse().expect("infallible"),
-            ))
-        } else {
-            // try to parse as a URI
-            Ok(AdvertisedAddress::Http(s.parse()?))
+        match s.trim() {
+            "" => Err(anyhow::anyhow!("Advertised address cannot be empty")),
+            address if address.starts_with("unix:") => {
+                parse_uds(&address[5..]).map(AdvertisedAddress::Uds)
+            }
+            address => parse_http(address),
+        }
+    }
+}
+
+fn parse_uds(s: &str) -> Result<PathBuf, Error> {
+    s.parse::<PathBuf>()
+        .with_context(|| format!("Failed to parse Unix domain socket path: '{}'", s))
+}
+
+fn parse_http(s: &str) -> Result<AdvertisedAddress, Error> {
+    let uri = s
+        .parse::<Uri>()
+        .with_context(|| format!("Invalid URI format: '{}'", s))?;
+
+    match uri.scheme_str() {
+        Some("http") | Some("https") => Ok(AdvertisedAddress::Http(uri)),
+        Some(other) => Err(anyhow::anyhow!("Unsupported URI scheme '{}'", other)),
+        None => Err(anyhow::anyhow!("Missing URI scheme in: '{}'", s)),
+    }
+}
+
+impl AdvertisedAddress {
+    /// Derives a `BindAddress` based on the advertised address
+    pub fn derive_bind_address(&self) -> BindAddress {
+        match self {
+            AdvertisedAddress::Http(uri) => {
+                let port = uri
+                    .authority()
+                    .and_then(|auth| auth.port_u16())
+                    .unwrap_or(5122); // Default to 5122 if no port is specified
```

Comment:
On second thought; I wonder if this is the least surprising possible behavior for operators?

If I specify the advertised address as `http://node-1` I might expect the bind address to be `0.0.0.0:80`. Similarly, then I would also expect `https://node-1` to default to 443 w/ TLS, which we don't currently support, but we can worry about TLS another day. 

Advertised address port being different from the bind port only makes sense if the operator has configured some additional proxy or port forwarding on the receiving node, in which case they should probably take control and explicitly set both binding and advertised address + port.

Maybe we should assert that the protocol is strictly `http` and default the port to 80, rather than 5122, if no explicit port is set in the URL? Any other takes from the team?

Reply:
IMO 
let's take the previous behavior (which is 0.0.0.0:5122 is the default if the bind address is not specified) as a reference, 
currently, we're sticking to that behavior if the port of the advertised address is not specified too 

so it doesn't feel that we changed much, we're falling back to the same bind address whenever it is not specified for whatever reason, I see this is okay as long as it is documented

Reply:
Sorry for the delay on this! I ran this past the team, and a couple of us feel strongly that this should be `80`. Since the property is a URL and it has a protocol component, we should respect the default well-known HTTP port of `80`[^1] if it's unspecified.

As a user of Restate, I'd be very surprised if I explicitly configured something as `http://example.com` and found it listening on a port that's not `80`. If you want to, you could add a check that asserts that `url.scheme()` is `"http"` - we don't support anything else at the moment.

```suggestion
                    .unwrap_or(80); // HTTP default port is 80 if unspecified
```

[^1]: https://datatracker.ietf.org/doc/html/rfc1340#page-9

Reply:
no worries at all, 
I have updated it, there are already tests to validate that the scheme is HTTP 
it's nice that you attached the footnote :D